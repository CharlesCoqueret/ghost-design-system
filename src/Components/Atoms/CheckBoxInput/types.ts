import { ReactElement } from 'react';

export interface IToggleEntry {
  value: string;
  checked?: boolean;
  label: string | ReactElement;
  highlighted?: boolean;
}
