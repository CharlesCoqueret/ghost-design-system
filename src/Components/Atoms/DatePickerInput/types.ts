export enum WeekDayEnum {
  SUNDAY = 0,
  MONDAY = 1,
  TUESDAY = 2,
  WEDNESDAY = 3,
  THURSDAY = 4,
  FRIDAY = 5,
  SATURDAY = 6,
}

export enum DateFormatEnum {
  MDY = 'MM/dd/yyyy',
  DMY = 'dd/MM/yyyy',
  YMD = 'yyyy/MM/dd',
  MMMddyyyy = 'MMM dd, yyyy',
  ddMMMyyyy = 'dd MMM yyyy',
}

export type DateFormat =
  | DateFormatEnum.MDY
  | DateFormatEnum.DMY
  | DateFormatEnum.YMD
  | DateFormatEnum.MMMddyyyy
  | DateFormatEnum.ddMMMyyyy
  | string;
