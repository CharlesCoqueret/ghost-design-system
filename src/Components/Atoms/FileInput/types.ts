export enum FileStatusEnum {
  /** File is being deleted, and should not be considered as available */
  DELETING = 'deleting',
  /** File is available */
  DONE = 'done',
  /** File is in error, and should not be considered as available */
  ERROR = 'error',
  /** File is being uploaded, and should not be considered as available */
  UPLOADING = 'uploading',
}

export interface IFile {
  /** Unique id (optional, if not provided, will be automatically generated) */
  uid?: string;
  /** Name of the file */
  name: string;
  /** Size of the file in bytes */
  size: number;
  /** Mime type of the file */
  type: string;
  /** Status of the file (optional, if not provided, will be automatically set to FileStatusEnum.DONE) */
  status?: FileStatusEnum;
  /** Progress of the upload, meaningfull while status is FileStatusEnum.UPLOADING */
  progress?: number;
  /** Error message of the upload, meaningfull while status is FileStatusEnum.ERROR */
  error?: string;
  /** Result provided by the upload call
   * (usually as a string, which might need to be parsed, for example using JSON.parse) */
  serverResponse?: unknown;
}
