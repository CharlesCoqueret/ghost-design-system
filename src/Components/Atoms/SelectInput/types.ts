export interface IOption {
  label: string;
  value: string;
  [key: string]: string;
}
