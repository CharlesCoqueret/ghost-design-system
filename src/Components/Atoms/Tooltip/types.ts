export enum MenuDirectionEnum {
  LEFT = 'left',
  RIGHT = 'right',
  TOP = 'top',
  BOTTOM = 'bottom',
}
