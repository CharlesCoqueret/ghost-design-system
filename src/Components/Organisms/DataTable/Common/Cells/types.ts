import {
  IColumnType,
  IExtraEditableDataTableProps,
  IExtraLineEditableInPlaceDataTableProps,
  IExtraStaticDataTableProps,
} from '../types';

export interface ICellProps<T, U extends IColumnType<T>> {
  column: U;
  dataTestId?: string;
  editing?: boolean;
  extra?: IExtraStaticDataTableProps<T> | IExtraLineEditableInPlaceDataTableProps<T> | IExtraEditableDataTableProps<T>;
  forcedValue?: T[keyof T] | number | string | Date | null;
  onChange?: (newValue: T[keyof T]) => void;
  row?: T;
  rowIndex: number;
}
