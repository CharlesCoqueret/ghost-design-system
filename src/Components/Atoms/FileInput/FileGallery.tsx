import React, { ReactElement, useState } from 'react';
import classnames from 'classnames';

import Button, { ColorButtonEnum } from '../../Molecules/Button/Button';
import { Icon } from '../Icon';
import { MenuDirectionEnum, Tooltip } from '../Tooltip';
import { FileStatusEnum, IFile } from './types';
import { formatBytes } from './fileUtils';

export interface IFileGallery {
  /** For test purpose only */
  dataTestId?: string;
  /** Disabled field (optional, default: false) */
  disabled?: boolean;
  /** File to be displayed */
  file: IFile;
  /** Reference to the progress map (optional, default: undefined) */
  progress?: Record<string, number>;
  /** Read only field (optional, default: false) */
  readOnly?: boolean;
  /** Show file size (optional, default: false) */
  showFileSize?: boolean;
  /** Show progress bar (optional, default: false) */
  showProgressBar?: boolean;
  /** Delete handler */
  updateFileDelete: (file: IFile) => Promise<void>;
  /** Download handler */
  updateFileDownload: (file: IFile) => Promise<void>;
  /** Localization of tooltip, and deletion action */
  localization?: {
    // Delete tooltip (optional, default: 'Delete')
    delete?: string;
    // Delete popover title (optional, default: 'Delete?')
    popoverConfirm?: string;
    // Delete popover cancel button (optional, default: 'Cancel')
    popoverCancel?: string;
    // Delete popover confirm button (optional, default: 'Confirm')
    popoverTitle?: string;
  };
}

const FileGallery = (props: IFileGallery): ReactElement => {
  const {
    dataTestId,
    disabled,
    file,
    localization,
    progress,
    readOnly,
    showFileSize,
    showProgressBar,
    updateFileDelete,
    updateFileDownload,
  } = props;

  const [downloading, setDownloading] = useState(false);

  const downloadable =
    file.status && ![FileStatusEnum.DELETING, FileStatusEnum.ERROR].includes(file.status) && !downloading;

  return (
    <Tooltip tooltip={file.error} direction={MenuDirectionEnum.TOP}>
      <div className={classnames('item-container', { error: file.error !== undefined })}>
        <div className='left'>
          <div className='paperclip-icon'>
            {file.error ? <Icon icon={['fal', 'exclamation-triangle']} /> : <Icon icon={['fal', 'paperclip']} />}
          </div>
          <div
            data-testid={dataTestId ? `${dataTestId}-download` : undefined}
            className={classnames('name', {
              downloadable: downloadable,
            })}
            onClick={
              downloadable
                ? () => {
                    setDownloading(true);
                    updateFileDownload(file)
                      .catch()
                      .finally(() => {
                        setDownloading(false);
                      });
                  }
                : undefined
            }>
            {file.name}
            {showFileSize && ` (${formatBytes(file.size)})`}
          </div>
          {downloading && (
            <Icon
              className='spinner-icon'
              data-testid={dataTestId ? `${dataTestId}-spinner` : undefined}
              icon={['fal', 'spinner']}
            />
          )}
        </div>
        <div className='right'>
          {showProgressBar && file.uid && progress && progress[file.uid] && (
            <progress className='progress' max={100} value={file.uid && progress[file.uid]} />
          )}

          {readOnly || disabled ? (
            <></>
          ) : file.status && [FileStatusEnum.UPLOADING, FileStatusEnum.DELETING].includes(file.status) ? (
            <Icon
              className='spinner-icon'
              data-testid={dataTestId ? `${dataTestId}-spinner` : undefined}
              icon={['fal', 'spinner']}
            />
          ) : (
            <Button
              className='delete-icon'
              icon={['fal', 'trash-alt']}
              color={ColorButtonEnum.REVERSED}
              dataTestId={dataTestId ? `${dataTestId}-delete` : undefined}
              onClick={
                // If the file is in error, we can delete without confirmation
                file.error
                  ? () => {
                      updateFileDelete(file);
                    }
                  : undefined
              }
              popover={
                // If the file is in error, we can delete without confirmation
                file.error
                  ? undefined
                  : {
                      title: localization?.popoverTitle ?? 'Delete?',
                      buttons: [
                        {
                          color: ColorButtonEnum.SECONDARY,
                          dataTestId: dataTestId ? `${dataTestId}-cancel` : undefined,
                          label: localization?.popoverCancel ?? 'Cancel',
                        },
                        {
                          color: ColorButtonEnum.PRIMARY,
                          dataTestId: dataTestId ? `${dataTestId}-confirm` : undefined,
                          label: localization?.popoverConfirm ?? 'Confirm',
                          onClick: () => {
                            updateFileDelete(file);
                          },
                        },
                      ],
                    }
              }
              tooltip={localization?.delete ?? 'Delete'}
            />
          )}
        </div>
      </div>
    </Tooltip>
  );
};

FileGallery.defaultProps = {
  dataTestId: undefined,
  disabled: false,
  progress: undefined,
  readOnly: false,
  showFileSize: false,
  showProgressBar: false,
};

export default FileGallery;
