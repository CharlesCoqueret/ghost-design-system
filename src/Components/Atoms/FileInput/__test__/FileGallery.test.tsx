import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import FileGallery from '../FileGallery';
import { FileStatusEnum } from '../types';

describe('FileGallery Component', () => {
  it('FileGallery renders and handles download and delete of done files', async () => {
    const updateFileDeleteMock = jest.fn().mockImplementation(async () => {
      return Promise.resolve();
    });
    const updateFileDownloadMock = jest.fn().mockImplementation(async () => {
      return Promise.resolve();
    });

    const file = {
      name: 'NAME',
      size: 1234,
      status: FileStatusEnum.DONE,
      type: 'TYPE',
      uid: '1',
    };

    const { container } = render(
      <FileGallery
        dataTestId='TEST-ID'
        file={file}
        updateFileDelete={updateFileDeleteMock}
        updateFileDownload={updateFileDownloadMock}
      />,
    );
    expect(container).toMatchSnapshot();

    const downloadButton = await screen.findByTestId('TEST-ID-download');

    userEvent.click(downloadButton);

    expect(updateFileDownloadMock).toBeCalledTimes(1);
    expect(updateFileDownloadMock).toBeCalledWith(file);

    const deleteButton = await screen.findByTestId('TEST-ID-delete');

    userEvent.click(deleteButton);

    const confirmButton = screen.getByTestId('TEST-ID-confirm');

    userEvent.click(confirmButton);

    expect(updateFileDeleteMock).toBeCalledTimes(1);
    expect(updateFileDeleteMock).toBeCalledWith(file);
  });

  it('FileGallery renders and handles delete of error files', async () => {
    const updateFileDeleteMock = jest.fn().mockImplementation(async () => {
      return Promise.resolve();
    });
    const updateFileDownloadMock = jest.fn().mockImplementation(async () => {
      return Promise.resolve();
    });

    const file = {
      error: 'ERROR',
      name: 'NAME',
      size: 1234,
      status: FileStatusEnum.ERROR,
      type: 'TYPE',
      uid: '1',
    };

    const { container } = render(
      <FileGallery
        dataTestId='TEST-ID'
        file={file}
        showFileSize
        updateFileDelete={updateFileDeleteMock}
        updateFileDownload={updateFileDownloadMock}
      />,
    );
    expect(container).toMatchSnapshot();

    const deleteButton = await screen.findByTestId('TEST-ID-delete');

    userEvent.click(deleteButton);

    expect(updateFileDeleteMock).toBeCalledTimes(1);
    expect(updateFileDeleteMock).toBeCalledWith(file);
  });

  it('FileGallery renders and handles delete of uploading files', async () => {
    const updateFileDeleteMock = jest.fn();
    const updateFileDownloadMock = jest.fn();

    const file = {
      name: 'NAME',
      size: 1234,
      status: FileStatusEnum.UPLOADING,
      type: 'TYPE',
      uid: '1',
    };

    const { container } = render(
      <FileGallery
        dataTestId='TEST-ID'
        file={file}
        progress={{ '1': 50 }}
        showProgressBar
        updateFileDelete={updateFileDeleteMock}
        updateFileDownload={updateFileDownloadMock}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('FileGallery renders readonly of uploading file', () => {
    const updateFileDeleteMock = jest.fn();
    const updateFileDownloadMock = jest.fn();

    const file = {
      name: 'NAME',
      size: 1234,
      status: FileStatusEnum.UPLOADING,
      type: 'TYPE',
      uid: '1',
    };

    const { container } = render(
      <FileGallery
        dataTestId='TEST-ID'
        file={file}
        progress={{ '1': 50 }}
        showProgressBar
        readOnly
        updateFileDelete={updateFileDeleteMock}
        updateFileDownload={updateFileDownloadMock}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('FileGallery renders disabled deleting file', () => {
    const updateFileDeleteMock = jest.fn();
    const updateFileDownloadMock = jest.fn();

    const file = {
      name: 'NAME',
      size: 1234,
      status: FileStatusEnum.DELETING,
      type: 'TYPE',
      uid: '1',
    };

    const { container } = render(
      <FileGallery
        dataTestId='TEST-ID'
        file={file}
        disabled
        updateFileDelete={updateFileDeleteMock}
        updateFileDownload={updateFileDownloadMock}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('FileGallery handles download without datatestid', async () => {
    const updateFileDeleteMock = jest.fn();
    const updateFileDownloadMock = jest.fn().mockImplementation(async () => {
      return Promise.resolve();
    });

    const file = {
      name: 'NAME',
      size: 1234,
      status: FileStatusEnum.DONE,
      type: 'TYPE',
      uid: '1',
    };

    const { container } = render(
      <FileGallery file={file} updateFileDelete={updateFileDeleteMock} updateFileDownload={updateFileDownloadMock} />,
    );
    expect(container).toMatchSnapshot();

    // Select download link
    userEvent.tab();
    userEvent.keyboard('{Enter}');

    expect(container).toMatchSnapshot();
    expect(updateFileDownloadMock).toBeCalledTimes(1);

    // Give time to updateFileDownloadMock to resolve
    await waitFor(async () => {
      await Promise.resolve();
    });
  });

  it('FileGallery handles delete without datatestid', async () => {
    const updateFileDeleteMock = jest.fn().mockImplementation(async () => {
      return Promise.resolve();
    });
    const updateFileDownloadMock = jest.fn();

    const file = {
      error: 'ERROR',
      name: 'NAME',
      size: 1234,
      status: FileStatusEnum.ERROR,
      type: 'TYPE',
      uid: '1',
    };

    const { container } = render(
      <FileGallery file={file} updateFileDelete={updateFileDeleteMock} updateFileDownload={updateFileDownloadMock} />,
    );

    expect(container).toMatchSnapshot();

    // Select first delete button
    userEvent.tab();
    userEvent.keyboard('{Enter}');

    expect(container).toMatchSnapshot();
    expect(updateFileDeleteMock).toBeCalledTimes(1);

    // Give time to updateFileDeleteMock to resolve
    await waitFor(async () => {
      await Promise.resolve();
    });
  });

  it('FileGallery handles uploading state without datatestid', async () => {
    const updateFileDeleteMock = jest.fn();
    const updateFileDownloadMock = jest.fn();

    const file = {
      name: 'NAME',
      size: 1234,
      status: FileStatusEnum.UPLOADING,
      type: 'TYPE',
      uid: '1',
    };

    const { container } = render(
      <FileGallery file={file} updateFileDelete={updateFileDeleteMock} updateFileDownload={updateFileDownloadMock} />,
    );

    expect(container).toMatchSnapshot();
  });
});
