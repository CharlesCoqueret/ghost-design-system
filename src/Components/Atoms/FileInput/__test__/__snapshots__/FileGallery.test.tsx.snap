// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileGallery Component FileGallery renders and handles delete of error files 1`] = `
<div>
  <span
    style="display: inline-block; width: 100%; overflow: hidden;"
  >
    <div
      class="item-container error"
    >
      <div
        class="left"
      >
        <div
          class="paperclip-icon"
        >
          <i
            class="icon"
            icon="fal,exclamation-triangle"
          />
        </div>
        <div
          class="name"
          data-testid="TEST-ID-download"
        >
          NAME
           (1.21 kB)
        </div>
      </div>
      <div
        class="right"
      >
        <div
          class="delete-icon"
          data-testid="TEST-ID-delete"
        >
          <i
            class="icon"
            icon="fal,trash-alt"
          />
        </div>
      </div>
    </div>
  </span>
  <div>
    <div
      class="gds-tooltip top"
      style="left: -99999px; top: -99999px; opacity: 0;"
    >
      <div
        class="tooltip-arrow"
      />
      <div
        class="tooltip-label"
      >
        ERROR
      </div>
    </div>
  </div>
</div>
`;

exports[`FileGallery Component FileGallery renders and handles delete of uploading files 1`] = `
<div>
  <div
    class="item-container"
  >
    <div
      class="left"
    >
      <div
        class="paperclip-icon"
      >
        <i
          class="icon"
          icon="fal,paperclip"
        />
      </div>
      <div
        class="name downloadable"
        data-testid="TEST-ID-download"
      >
        NAME
      </div>
    </div>
    <div
      class="right"
    >
      <progress
        class="progress"
        max="100"
        value="50"
      />
      <svg
        aria-hidden="true"
        class="icon gds-loader size-1x spinner-icon"
        data-testid="TEST-ID-spinner"
        focusable="false"
        height="50"
        icon="fal,spinner"
        preserveAspectRatio="xMidYMid"
        role="img"
        style="background: transparent;"
        viewBox="0 0 100 100"
        width="50"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="50"
          cy="50"
          fill="transparent"
          r="30"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-width="10"
        >
          <animatetransform
            attributeName="transform"
            dur="2s"
            keyTimes="0;0.5;1"
            repeatCount="indefinite"
            type="rotate"
            values="0 50 50;180 50 50;720 50 50"
          />
          <animate
            attributeName="stroke-dasharray"
            dur="2s"
            keyTimes="0;0.5;1"
            repeatCount="indefinite"
            values="20 150; 150 20; 20 150"
          />
        </circle>
      </svg>
    </div>
  </div>
</div>
`;

exports[`FileGallery Component FileGallery renders and handles download and delete of done files 1`] = `
<div>
  <div
    class="item-container"
  >
    <div
      class="left"
    >
      <div
        class="paperclip-icon"
      >
        <i
          class="icon"
          icon="fal,paperclip"
        />
      </div>
      <div
        class="name downloadable"
        data-testid="TEST-ID-download"
      >
        NAME
      </div>
    </div>
    <div
      class="right"
    >
      <div
        class="delete-icon"
        data-testid="TEST-ID-delete"
      >
        <i
          class="icon"
          icon="fal,trash-alt"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`FileGallery Component FileGallery renders disabled deleting file 1`] = `
<div>
  <div
    class="item-container"
  >
    <div
      class="left"
    >
      <div
        class="paperclip-icon"
      >
        <i
          class="icon"
          icon="fal,paperclip"
        />
      </div>
      <div
        class="name"
        data-testid="TEST-ID-download"
      >
        NAME
      </div>
    </div>
    <div
      class="right"
    />
  </div>
</div>
`;

exports[`FileGallery Component FileGallery renders readonly of uploading file 1`] = `
<div>
  <div
    class="item-container"
  >
    <div
      class="left"
    >
      <div
        class="paperclip-icon"
      >
        <i
          class="icon"
          icon="fal,paperclip"
        />
      </div>
      <div
        class="name downloadable"
        data-testid="TEST-ID-download"
      >
        NAME
      </div>
    </div>
    <div
      class="right"
    >
      <progress
        class="progress"
        max="100"
        value="50"
      />
    </div>
  </div>
</div>
`;

exports[`FileGallery Component FileGallery renders without datatestid 1`] = `
<div>
  <div
    class="item-container"
  >
    <div
      class="left"
    >
      <div
        class="paperclip-icon"
      >
        <i
          class="icon"
          icon="fal,paperclip"
        />
      </div>
      <div
        class="name downloadable"
      >
        NAME
      </div>
    </div>
    <div
      class="right"
    >
      <div
        class="delete-icon"
      >
        <i
          class="icon"
          icon="fal,trash-alt"
        />
      </div>
    </div>
  </div>
</div>
`;
